src/Covenant/Type.hs

```diff
@@ -588,3 +656,136 @@ returnHelper xs = case NonEmpty.uncons xs of
     if Vector.length ys == 0
       then pure y
       else Nothing
+
+unify ::
+  ValT Renamed ->
+  ValT Renamed ->
+  Either TypeAppError (Map (Index "tyvar") (ValT Renamed))
+unify expected actual =
+  catchError
+    ( case expected of
+        Abstraction t1 -> case t1 of
+          -- Unifiables unify with everything, and require a substitutional rewrite.
+          Unifiable index1 -> pure . Map.singleton index1 $ actual
+          Rigid level1 index1 -> expectRigid level1 index1
+          Wildcard scopeId1 index1 -> expectWildcard scopeId1 index1
```

Comment:
Shouldn't a wildcard unify with anything? :thinking: e.g. also with concrete type `BuiltinFlat IntegerT`? :thinking:

Reply:
Hmm, I don't follow why not. Could you explain further :thinking:

Reply:
Consider what the type `(forall a . a -> !Integer) -> !Integer` says. Roughly-speaking, we want a thunk that can take _any type whatsoever_ and turn it into an `Integer`. Therefore, the `a` wildcard there can _only_ unify with a unifiable: if we were to try and pass `(Integer -> !Integer)` as an argument to this function, it _should_ be unsuitable, as `(Integer -> !Integer)` cannot take _any type whatsoever_ and turn it into an integer.